Could not find BigQuery jobs in not US location

I'm trying to throw a query to BigQuery's dataset in asia-northeast1, but digdag can't find job's status.
It seem to be searching jobs with format ${PROJECT_NAME}:${JOB_ID} instead of ${PROJECT_NAME}:${LOCATION}:${JOB_ID}. isn't it?

### error logs
```
2019-04-16 21:41:43 +0900 [ERROR] (0095@[0:dataFlow]+main+test) io.digdag.core.agent.OperatorManager: Task +main+test failed.
digdag_1       | 404 Not Found
digdag_1       | {
digdag_1       |   "code" : 404,
digdag_1       |   "errors" : [ {
digdag_1       |     "domain" : "global",
digdag_1       |     "message" : "Not found: Job my-gcp-project:digdag_s0_p_my-digdag-project_w_main_t_950_a_13_6aa3ad0c-ebd2-4f78-a6b7-0c3e47957436",
digdag_1       |     "reason" : "notFound"
digdag_1       |   } ],
digdag_1       |   "message" : "Not found: Job my-gcp-project:digdag_s0_p_my-digdag-project_w_main_t_950_a_13_6aa3ad0c-ebd2-4f78-a6b7-0c3e47957436",
digdag_1       |   "status" : "NOT_FOUND"
digdag_1       | } (google json response)
```

### main.dig
```
test:
   bq>: queries/test.sql
   location: asia-northeast1
   dataset: my_dataset
```

### queries/test.sql
```
SELECT
  *
FROM
  `my_dataset.my_table`
```

--- a/digdag-docs/src/operators/bq.md
+++ b/digdag-docs/src/operators/bq.md
@@ -46,6 +46,16 @@ When you set those parameters, use [digdag secrets command](https://docs.digdag.
   bq>: queries/step1.sql
   ```
 
+* **location**: LOCATION
+
+  The location where the query job should run. See [locations](https://cloud.google.com/bigquery/docs/locations) for a list of available locations.
+
+  Examples:
+
+  ```
+  location: asia-northeast1
+  ```
+
 * **dataset**: NAME
 
   Specifies the default dataset to use in the query and in the `destination_table` parameter.
@@ -164,4 +174,3 @@ When you set those parameters, use [digdag secrets command](https://docs.digdag.
   The id of the BigQuery job that executed this query.
 
   Note: `bq.last_jobid` parameter is kept only for backward compatibility but you must not use it because it will be removed removed in a near future release.
-

--- a/digdag-docs/src/operators/bq_extract.md
+++ b/digdag-docs/src/operators/bq_extract.md
@@ -57,6 +57,15 @@ When you set those parameters, use [digdag secrets command](https://docs.digdag.
     - gs://my_bucket/my_export_2.csv
   ```
 
+* **location**: LOCATION
+  The location where the job should run. The table and the destination must be in this location. See [BigQuery locations](https://cloud.google.com/bigquery/docs/locations) for a list of available locations.
+
+  Examples:
+
+  ```
+  location: asia-northeast1
+  ```
+
 * **print_header**: BOOLEAN
   Whether to print out a header row in the results. *Default*: `true`.
 
@@ -106,4 +115,3 @@ When you set those parameters, use [digdag secrets command](https://docs.digdag.
   The id of the BigQuery job that performed this export.
 
   Note: `bq.last_jobid` parameter is kept only for backward compatibility but you must not use it because it will be removed removed in a near future release.
-

--- a/digdag-docs/src/operators/bq_load.md
+++ b/digdag-docs/src/operators/bq_load.md
@@ -80,6 +80,16 @@ When you set those parameters, use [digdag secrets command](https://docs.digdag.
   destination_table: some_dataset.some_partitioned_table$20160101
   ```
 
+* **location**: LOCATION
+
+  The location where the job should run. The source GCS bucket and the table must be in this location. See [BigQuery locations](https://cloud.google.com/bigquery/docs/locations) for a list of available locations.
+
+  Examples:
+
+  ```
+  location: asia-northeast1
+  ```
+
 * **project**: NAME
 
   The project that the table is located in or should be created in. Can also be specified directly in the table reference or the dataset parameter.
@@ -280,4 +290,3 @@ When you set those parameters, use [digdag secrets command](https://docs.digdag.
   The id of the BigQuery job that performed this import.
 
   Note: `bq.last_jobid` parameter is kept only for backward compatibility but you must not use it because it will be removed removed in a near future release.
-

--- a/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BaseBqJobOperator.java
+++ b/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BaseBqJobOperator.java
@@ -22,7 +22,8 @@ abstract class BaseBqJobOperator
     protected TaskResult run(BqClient bq, String projectId)
     {
         BqJobRunner jobRunner = new BqJobRunner(request, bq, projectId);
-        Job completed = jobRunner.runJob(jobConfiguration(projectId));
+        String location = params.getOptional("location", String.class).orNull();
+        Job completed = jobRunner.runJob(jobConfiguration(projectId), location);
         return result(completed);
     }
 

--- a/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BaseBqJobOperator.java
+++ b/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BaseBqJobOperator.java
@@ -2,6 +2,7 @@ package io.digdag.standards.operator.gcp;
 
 import com.google.api.services.bigquery.model.Job;
 import com.google.api.services.bigquery.model.JobConfiguration;
+import com.google.common.base.Optional;
 import io.digdag.client.config.Config;
 import io.digdag.client.config.ConfigFactory;
 import io.digdag.client.config.ConfigKey;
@@ -22,7 +23,7 @@ abstract class BaseBqJobOperator
     protected TaskResult run(BqClient bq, String projectId)
     {
         BqJobRunner jobRunner = new BqJobRunner(request, bq, projectId);
-        String location = params.getOptional("location", String.class).orNull();
+        Optional<String> location = params.getOptional("location", String.class);
         Job completed = jobRunner.runJob(jobConfiguration(projectId), location);
         return result(completed);
     }

--- a/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqClient.java
+++ b/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqClient.java
@@ -170,11 +170,13 @@ class BqClient
                 .execute();
     }
 
-    Job jobStatus(String projectId, String jobId)
+    Job jobStatus(String projectId, String jobId, String location)
             throws IOException
     {
         return client.jobs()
                 .get(projectId, jobId)
+                // newer version of google-api-services-bigquery-v2 has setLocation()
+                .set("location", location)
                 .execute();
     }
 

--- a/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqClient.java
+++ b/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqClient.java
@@ -170,14 +170,21 @@ class BqClient
                 .execute();
     }
 
-    Job jobStatus(String projectId, String jobId, String location)
+    Job jobStatus(String projectId, String jobId, Optional<String> location)
             throws IOException
     {
-        return client.jobs()
-                .get(projectId, jobId)
-                // newer version of google-api-services-bigquery-v2 has setLocation()
-                .set("location", location)
-                .execute();
+        if (location.isPresent()) {
+            return client.jobs()
+                    .get(projectId, jobId)
+                    // newer version of google-api-services-bigquery-v2 has setLocation()
+                    .set("location", location.get())
+                    .execute();
+        }
+        else {
+            return client.jobs()
+                    .get(projectId, jobId)
+                    .execute();
+        }
     }
 
     static class Factory

--- a/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqJobRunner.java
+++ b/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqJobRunner.java
@@ -49,7 +49,7 @@ class BqJobRunner
         this.projectId = Objects.requireNonNull(projectId, "projectId");
     }
 
-    Job runJob(JobConfiguration config)
+    Job runJob(JobConfiguration config, String location)
     {
         // Generate job id
         Optional<String> jobId = state.params().getOptional(JOB_ID, String.class);
@@ -97,7 +97,7 @@ class BqJobRunner
                             .withErrorMessage("BigQuery job status check failed: %s", canonicalJobId)
                             .run(s -> {
                                 logger.info("Checking BigQuery job status: {}", canonicalJobId);
-                                return bq.jobStatus(projectId, jobId.get());
+                                return bq.jobStatus(projectId, jobId.get(), location);
                             });
 
                     // Done yet?

--- a/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqJobRunner.java
+++ b/digdag-standards/src/main/java/io/digdag/standards/operator/gcp/BqJobRunner.java
@@ -49,7 +49,7 @@ class BqJobRunner
         this.projectId = Objects.requireNonNull(projectId, "projectId");
     }
 
-    Job runJob(JobConfiguration config, String location)
+    Job runJob(JobConfiguration config, Optional<String> location)
     {
         // Generate job id
         Optional<String> jobId = state.params().getOptional(JOB_ID, String.class);

--- a/digdag-tests/src/test/java/acceptance/td/BigQueryIT.java
+++ b/digdag-tests/src/test/java/acceptance/td/BigQueryIT.java
@@ -48,7 +48,17 @@ import java.nio.file.Path;
 import java.time.Duration;
 import java.util.concurrent.ConcurrentHashMap;
 
-import static acceptance.td.GcpUtil.*;
+import static acceptance.td.GcpUtil.BQ_TAG;
+import static acceptance.td.GcpUtil.GCP_CREDENTIAL;
+import static acceptance.td.GcpUtil.GCS_PREFIX;
+import static acceptance.td.GcpUtil.GCS_TEST_BUCKET;
+import static acceptance.td.GcpUtil.GCS_TEST_BUCKET_ASIA;
+import static acceptance.td.GcpUtil.createDataset;
+import static acceptance.td.GcpUtil.createTable;
+import static acceptance.td.GcpUtil.datasetExists;
+import static acceptance.td.GcpUtil.getDatasetLocation;
+import static acceptance.td.GcpUtil.listTables;
+import static acceptance.td.GcpUtil.tableExists;
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.is;

--- a/digdag-tests/src/test/java/acceptance/td/BigQueryIT.java
+++ b/digdag-tests/src/test/java/acceptance/td/BigQueryIT.java
@@ -48,15 +48,7 @@ import java.nio.file.Path;
 import java.time.Duration;
 import java.util.concurrent.ConcurrentHashMap;
 
-import static acceptance.td.GcpUtil.BQ_TAG;
-import static acceptance.td.GcpUtil.GCP_CREDENTIAL;
-import static acceptance.td.GcpUtil.GCS_PREFIX;
-import static acceptance.td.GcpUtil.GCS_TEST_BUCKET;
-import static acceptance.td.GcpUtil.createDataset;
-import static acceptance.td.GcpUtil.createTable;
-import static acceptance.td.GcpUtil.datasetExists;
-import static acceptance.td.GcpUtil.listTables;
-import static acceptance.td.GcpUtil.tableExists;
+import static acceptance.td.GcpUtil.*;
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.is;
@@ -220,8 +212,36 @@ public class BigQueryIT
         public void testQuery()
                 throws Exception
         {
+            String testQueryTokyoDataset1 = BQ_TAG + "_query_tokyo_dataset_1";
+            String testQueryTable1 = "test_query_table1";
+
+            createDataset(bq, gcpProjectId, testQueryTokyoDataset1, "asia-northeast1");
+            createTable(bq, gcpProjectId, testQueryTokyoDataset1, new Table()
+                    .setTableReference(new TableReference()
+                            .setProjectId(gcpProjectId)
+                            .setDatasetId(testQueryTokyoDataset1)
+                            .setTableId(testQueryTable1))
+                    .setSchema(new TableSchema()
+                            .setFields(ImmutableList.of(
+                                    new TableFieldSchema().setName("f1").setType("STRING"),
+                                    new TableFieldSchema().setName("f2").setType("STRING")
+                            ))));
+
+            retryExecutor.run(() ->
+                    bq.tabledata().insertAll(gcpProjectId, testQueryTokyoDataset1, testQueryTable1, new TableDataInsertAllRequest()
+                            .setRows(ImmutableList.of(
+                                            new TableDataInsertAllRequest.Rows().setJson(ImmutableMap.of("f1", "v1a", "f2", "v2a")),
+                                            new TableDataInsertAllRequest.Rows().setJson(ImmutableMap.of("f1", "v1b", "f2", "v2b"))
+                                    )
+                            )
+                    )
+            );
+
             addWorkflow(projectDir, "acceptance/bigquery/query.dig");
-            Id attemptId = pushAndStart(server.endpoint(), projectDir, "query", ImmutableMap.of("outfile", outfile.toString()));
+            Id attemptId = pushAndStart(server.endpoint(), projectDir, "query", ImmutableMap.of(
+                    "outfile", outfile.toString(),
+                    "test_query_tokyo_dataset_1", testQueryTokyoDataset1,
+                    "test_query_table_1", testQueryTable1));
             expect(Duration.ofMinutes(5), attemptSuccess(server.endpoint(), attemptId));
             assertThat(Files.exists(outfile), is(true));
         }
@@ -249,11 +269,7 @@ public class BigQueryIT
 
             // Create output dataset
             String datasetId = BQ_TAG + "_load_test";
-            Dataset dataset = new Dataset().setDatasetReference(new DatasetReference()
-                    .setProjectId(gcpProjectId)
-                    .setDatasetId(datasetId));
-            retryExecutor.run(() -> bq.datasets().insert(gcpProjectId, dataset)
-                    .execute());
+            createDataset(bq, gcpProjectId, datasetId);
 
             // Run load
             String tableId = "data";
@@ -285,11 +301,7 @@ public class BigQueryIT
             // Create source table
             String tableId = "data";
             String datasetId = BQ_TAG + "_extract_test";
-            Dataset dataset = new Dataset().setDatasetReference(new DatasetReference()
-                    .setProjectId(gcpProjectId)
-                    .setDatasetId(datasetId));
-            retryExecutor.run(() -> bq.datasets().insert(gcpProjectId, dataset)
-                    .execute());
+            createDataset(bq, gcpProjectId, datasetId);
             Table table = new Table().setTableReference(new TableReference()
                     .setProjectId(gcpProjectId)
                     .setTableId(tableId))
@@ -363,11 +375,13 @@ public class BigQueryIT
             String testCreateTable2EmptyDataset = BQ_TAG + "_create_table_2_empty_dataset";
             String testCreateTable3CreateDataset = BQ_TAG + "_create_table_3_create_dataset";
             String testCreateTable4ExistingDataset = BQ_TAG + "_create_table_4_existing_dataset";
+            String testCreateTableAtTokyo1CreateDataset = BQ_TAG + "_create_table_at_tokyo_1_create_dataset";
             String testCreateTable1 = "test_create_table_1";
             String testCreateTable2 = "test_create_table_2";
             String testCreateTable3 = "test_create_table_3";
             String testCreateTable4 = "test_create_table_4";
             String testCreateTable5 = "test_create_table_5";
+            String testCreateTableAtTokyo1 = "test_create_table_at_tokyo_1";
             String testDeleteTable1 = "test_delete_table_1";
             String testDeleteTable2 = "test_delete_table_2";
             String testEmptyTable2EmptyDataset = BQ_TAG + "_empty_table_2_empty_dataset";
@@ -426,9 +440,11 @@ public class BigQueryIT
                     .put("test_create_table_3", testCreateTable3)
                     .put("test_create_table_4", testCreateTable4)
                     .put("test_create_table_5", testCreateTable5)
+                    .put("test_create_table_at_tokyo_1", testCreateTableAtTokyo1)
                     .put("test_create_table_2_empty_dataset", testCreateTable2EmptyDataset)
                     .put("test_create_table_3_create_dataset", testCreateTable3CreateDataset)
                     .put("test_create_table_4_existing_dataset", testCreateTable4ExistingDataset)
+                    .put("test_create_table_at_tokyo_1_create_dataset", testCreateTableAtTokyo1CreateDataset)
                     .put("test_delete_table_2_dataset", testDeleteTable2ExistingDataset)
                     .put("test_delete_table_1", testDeleteTable1)
                     .put("test_delete_table_2", testDeleteTable2)
@@ -453,6 +469,7 @@ public class BigQueryIT
             assertThat(datasetExists(bq, gcpProjectId, testEmptyDataset2), is(true));
             assertThat(datasetExists(bq, gcpProjectId, testDeleteDataset1), is(false));
             assertThat(datasetExists(bq, gcpProjectId, testDeleteDataset2), is(false));
+            assertThat(getDatasetLocation(bq, gcpProjectId, testCreateTableAtTokyo1CreateDataset), is(Optional.of("asia-northeast1")));
 
             assertThat(listTables(bq, gcpProjectId, testEmptyDataset1), is(empty()));
             assertThat(listTables(bq, gcpProjectId, testEmptyDataset2), is(empty()));
@@ -462,6 +479,7 @@ public class BigQueryIT
             assertThat(tableExists(bq, gcpProjectId, testCreateTable3CreateDataset, testCreateTable3), is(true));
             assertThat(tableExists(bq, gcpProjectId, testCreateTable4ExistingDataset, testCreateTable4), is(true));
             assertThat(tableExists(bq, gcpProjectId, testDefaultDataset, testCreateTable5), is(true));
+            assertThat(tableExists(bq, gcpProjectId, testCreateTableAtTokyo1CreateDataset, testCreateTableAtTokyo1), is(true));
 
             assertThat(tableExists(bq, gcpProjectId, testDefaultDataset, testDeleteTable1), is(false));
             assertThat(tableExists(bq, gcpProjectId, testDeleteTable2ExistingDataset, testDeleteTable2), is(false));

--- a/digdag-tests/src/test/java/acceptance/td/BigQueryIT.java
+++ b/digdag-tests/src/test/java/acceptance/td/BigQueryIT.java
@@ -287,6 +287,43 @@ public class BigQueryIT
             Table destinationTable = retryExecutor.run(() -> bq.tables().get(gcpProjectId, datasetId, tableId).execute());
             assertThat(destinationTable.getTableReference().getTableId(), is(tableId));
         }
+
+        @Test
+        public void testLoadAsia()
+                throws Exception
+        {
+            assertThat(GCS_TEST_BUCKET_ASIA, not(isEmptyOrNullString()));
+
+            // Create source data object
+            String objectName = GCS_PREFIX + "test.csv";
+            byte[] data = Joiner.on('\n').join("a,b", "c,d").getBytes(UTF_8);
+            InputStreamContent content = new InputStreamContent("text/csv", new ByteArrayInputStream(data))
+                    .setLength(data.length);
+            StorageObject metadata = new StorageObject().setName(objectName);
+            retryExecutor.run(() -> gcs.objects()
+                    .insert(GCS_TEST_BUCKET_ASIA, metadata, content)
+                    .execute());
+
+            // Create output dataset
+            String datasetId = BQ_TAG + "_load_test_asia";
+            createDataset(bq, gcpProjectId, datasetId, "asia-northeast1");
+
+            // Run load
+            String tableId = "data";
+            addWorkflow(projectDir, "acceptance/bigquery/load-asia.dig");
+            Id attemptId = pushAndStart(server.endpoint(), projectDir, "load-asia", ImmutableMap.of(
+                    "source_bucket", GCS_TEST_BUCKET_ASIA,
+                    "source_object", objectName,
+                    "target_dataset", datasetId,
+                    "target_table", tableId,
+                    "outfile", outfile.toString()));
+            expect(Duration.ofMinutes(5), attemptSuccess(server.endpoint(), attemptId));
+            assertThat(Files.exists(outfile), is(true));
+
+            // Check that destination table was created
+            Table destinationTable = retryExecutor.run(() -> bq.tables().get(gcpProjectId, datasetId, tableId).execute());
+            assertThat(destinationTable.getTableReference().getTableId(), is(tableId));
+        }
     }
 
     public static class ExtractIT
@@ -345,6 +382,67 @@ public class BigQueryIT
             retryExecutor.run(() -> {
                 try {
                     gcs.objects().get(GCS_TEST_BUCKET, objectName)
+                            .executeMediaAndDownloadTo(data);
+                }
+                catch (IOException e) {
+                    throw ThrowablesUtil.propagate(e);
+                }
+            });
+        }
+
+        @Test
+        public void testExtractAsia()
+                throws Exception
+        {
+            assertThat(GCS_TEST_BUCKET_ASIA, not(isEmptyOrNullString()));
+
+            // Create source table
+            String tableId = "data";
+            String datasetId = BQ_TAG + "_extract_test_asia";
+            createDataset(bq, gcpProjectId, datasetId, "asia-northeast1");
+            Table table = new Table().setTableReference(new TableReference()
+                            .setProjectId(gcpProjectId)
+                            .setTableId(tableId))
+                    .setSchema(new TableSchema()
+                            .setFields(ImmutableList.of(
+                                    new TableFieldSchema().setName("foo").setType("STRING"),
+                                    new TableFieldSchema().setName("bar").setType("STRING")
+                            )));
+            retryExecutor.run(() -> bq.tables().insert(gcpProjectId, datasetId, table)
+                    .execute());
+
+            // Populate source table
+            TableDataInsertAllRequest content = new TableDataInsertAllRequest()
+                    .setRows(ImmutableList.of(
+                            new TableDataInsertAllRequest.Rows().setJson(ImmutableMap.of(
+                                    "foo", "a",
+                                    "bar", "b")),
+                            new TableDataInsertAllRequest.Rows().setJson(ImmutableMap.of(
+                                    "foo", "c",
+                                    "bar", "d"))));
+            retryExecutor.run(() -> bq.tabledata().insertAll(gcpProjectId, datasetId, tableId, content)
+                    .execute());
+
+            // Run extract
+            String objectName = GCS_PREFIX + "test.csv";
+            addWorkflow(projectDir, "acceptance/bigquery/extract-asia.dig");
+            Id attemptId = pushAndStart(server.endpoint(), projectDir, "extract-asia", ImmutableMap.of(
+                    "src_dataset", datasetId,
+                    "src_table", tableId,
+                    "dst_bucket", GCS_TEST_BUCKET_ASIA,
+                    "dst_object", objectName,
+                    "outfile", outfile.toString()));
+            expect(Duration.ofMinutes(5), attemptSuccess(server.endpoint(), attemptId));
+            assertThat(Files.exists(outfile), is(true));
+
+            // Check that destination file was created
+            StorageObject metadata = retryExecutor.run(() -> gcs.objects().get(GCS_TEST_BUCKET_ASIA, objectName)
+                    .execute());
+            assertThat(metadata.getName(), is(objectName));
+            ByteArrayOutputStream data = new ByteArrayOutputStream();
+            retryExecutor.run(() -> {
+                try {
+                    gcs.objects().get(GCS_TEST_BUCKET_ASIA, objectName)
                             .executeMediaAndDownloadTo(data);
                 }
                 catch (IOException e) {

--- a/digdag-tests/src/test/java/acceptance/td/GcpUtil.java
+++ b/digdag-tests/src/test/java/acceptance/td/GcpUtil.java
@@ -36,6 +36,7 @@ public class GcpUtil
 
     static final String GCP_CREDENTIAL = System.getenv().getOrDefault("GCP_CREDENTIAL", "");
     static final String GCS_TEST_BUCKET = System.getenv().getOrDefault("GCS_TEST_BUCKET", "");
+    static final String GCS_TEST_BUCKET_ASIA = System.getenv().getOrDefault("GCS_TEST_BUCKET_ASIA", "");
 
     static final String GCP_PROJECT_ID;
 

--- a/digdag-tests/src/test/java/acceptance/td/GcpUtil.java
+++ b/digdag-tests/src/test/java/acceptance/td/GcpUtil.java
@@ -13,6 +13,7 @@ import com.google.api.services.bigquery.model.TableReference;
 import com.google.api.services.storage.Storage;
 import com.google.api.services.storage.model.Objects;
 import com.google.api.services.storage.model.StorageObject;
+import com.google.common.base.Optional;
 import io.digdag.client.DigdagClient;
 import io.digdag.commons.ThrowablesUtil;
 import io.digdag.util.RetryExecutor;
@@ -138,6 +139,18 @@ public class GcpUtil
         return bq.tables().insert(projectId, datasetId, table).execute();
     }
 
+    static Dataset createDataset(Bigquery bq, String projectId, String datasetId, String location)
+            throws IOException, RetryExecutor.RetryGiveupException
+    {
+        Dataset dataset = new Dataset()
+                .setDatasetReference(new DatasetReference()
+                        .setDatasetId(datasetId))
+                .setLocation(location);
+        Dataset created = createDataset(bq, projectId, dataset);
+        assertThat(datasetExists(bq, projectId, datasetId), is(true));
+        return created;
+    }
+
     static Dataset createDataset(Bigquery bq, String projectId, String datasetId)
             throws IOException, RetryExecutor.RetryGiveupException
     {
@@ -168,6 +181,13 @@ public class GcpUtil
             }
             throw e;
         }
+    }
+
+    static Optional<String> getDatasetLocation(Bigquery bq, String projectId, String datasetId)
+        throws IOException
+    {
+            Dataset dataset = bq.datasets().get(projectId, datasetId).execute();
+            return Optional.fromNullable(dataset.getLocation());
     }
 
     static List<TableList.Tables> listTables(Bigquery bq, String projectId, String datasetId)

--- a/digdag-tests/src/test/resources/acceptance/bigquery/ddl.dig
+++ b/digdag-tests/src/test/resources/acceptance/bigquery/ddl.dig
@@ -15,6 +15,8 @@ timezone: UTC
         l2: v2
     - ${test_create_table_3_create_dataset}
     - ${test_empty_table_3_create_dataset}
+    - id: ${test_create_table_at_tokyo_1_create_dataset}
+      location: asia-northeast1
 
   delete_datasets:
     - ${test_delete_dataset_1}
@@ -43,6 +45,7 @@ timezone: UTC
         fields:
           - {name: f1, type: STRING}
           - {name: f2, type: STRING}
+    - ${test_create_table_at_tokyo_1_create_dataset}.${test_create_table_at_tokyo_1}
 
   delete_tables:
     - ${test_delete_table_1}

--- a/digdag-tests/src/test/resources/acceptance/bigquery/extract-asia.dig
+++ b/digdag-tests/src/test/resources/acceptance/bigquery/extract-asia.dig
@@ -1,0 +1,9 @@
+timezone: UTC
+
++run:
+  bq_extract>: ${src_dataset}.${src_table}
+  destination: gs://${dst_bucket}/${dst_object}
+  location: asia-northeast1
+
++post:
+  sh>: touch ${outfile}

--- a/digdag-tests/src/test/resources/acceptance/bigquery/load-asia.dig
+++ b/digdag-tests/src/test/resources/acceptance/bigquery/load-asia.dig
@@ -1,0 +1,15 @@
+timezone: UTC
+
++run:
+  bq_load>: gs://${source_bucket}/${source_object}
+  schema:
+    fields:
+      - name: foo
+        type: STRING
+      - name: bar
+        type: STRING
+  destination_table: ${target_dataset}.${target_table}
+  location: asia-northeast1
+
++post:
+  sh>: touch ${outfile}

--- a/digdag-tests/src/test/resources/acceptance/bigquery/query.dig
+++ b/digdag-tests/src/test/resources/acceptance/bigquery/query.dig
@@ -10,5 +10,11 @@ timezone: UTC
       ORDER BY weight_pounds DESC LIMIT 10;
   dataset: publicdata:samples
 
++run_tokyo1:
+  bq>:
+    data: SELECT f1, f2 FROM ${test_query_table_1};
+  dataset: ${test_query_tokyo_dataset_1}
+  location: asia-northeast1
+
 +post:
   sh>: touch ${outfile}

